STOR-3 feature products

Ktechen opened the pull request:
Add Model
Add DB dependency and application properties
Add Repository Interface to connection with Storage

Ktechen committed:
STOR-8 Create Address.java and Product.java Model

Ktechen committed:
Add dependency for persistence

Ktechen committed:
STOR-8 refactor to model dir

Ktechen committed:
STOR-8 refactor rename Model to model

Ktechen committed:
STOR-7 Add ProductRepository and dependency

Ktechen committed:
STOR-8 Add Column parameter

Ktechen committed:
Merge branch 'main' of https://github.com/But-Better/StorageApi into STOR-3-Feature-Products

Ktechen committed:
STOR-8 Add Column parameter

Ktechen committed:
STOR-8 add and change comments

Ktechen committed:
STOR-8 Add ProductInformation and refactor Product

Ktechen committed:
STOR-8 Remove Product.java is part of applicationService
And added comment to ProductInformation

Ktechen committed:
STOR-7 Change product to productinformation

Ktechen committed:
STOR-7 rename ProductRepository.java to IProductRepository.java

Ktechen committed:
STOR-7 Add repo methodes

Ktechen committed:
STOR-7 remove ImportServices.java is wrong STOR

Ktechen committed:
STOR-20 Add .env value to application.properties

Ktechen committed:
STOR-20 add Postgre and update hibernate properties

Ktechen committed:
STOR-8 UUID add insertable = false, updatable = false and GenerationType.AUTO

Ktechen committed:
STOR-7 update hashcode() if company name null

MathisHTW commented on src/main/java/com/butbetter/storage/model/Address.java:
```diff
@@ -0,0 +1,142 @@
+        return "Address{" +
+                "uuid=" + uuid +
+                ", name='" + name + '\'' +
+                ", companyName='" + companyName + '\'' +
```
denke es wäre besser nichts auszugeben wenn kein companyName existiert aber ist mir egal
+ Objects.isNull(companyName)? "" + ", companyName='" + companyName + '\''+

Ktechen replied:
Null ist doch okay bei toString()

Commit:
STOR-21 add test for hash code generation on potential null value

MathisHTW commented on src/main/java/com/butbetter/storage/model/ProductInformation.java:
```diff
@@ -0,0 +1,100 @@
+    @Column(name = "delivery_time", nullable = false)
+    private Date deliveryTime;
+
+    @Column(name = "amount", nullable = false)
```
dont allow negative values for amount 
-> use @Min(0)
https://spring.io/guides/gs/validating-form-input/

maybe we should also consider having a top value so a overflow cannot happen
in that case we might use
@Range(min= 0, max=100000)

https://stackoverflow.com/questions/44007492/enforce-constraints-on-value-annotated-field-in-spring-boot-application

Ktechen replied:
I could add a range to amount

MathisHTW replied:
min and max values are also part of validation in the service

MathisHTW commented on src/main/java/com/butbetter/storage/model/Address.java:
```diff
@@ -0,0 +1,142 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "id", insertable = false, updatable = false, nullable = false)
```
please use @NotNull instead of using nullable = false.
@NotNull will not trigger the underlying db but nullable = false will.
-> @NotNull is better performing

https://www.baeldung.com/hibernate-notnull-vs-nullable

Ktechen replied:
NotNull is not a Part of hibernate

Ktechen replied:
I dont see any diffent between NotNull and nullable = false

MathisHTW replied:
Even though both of them prevent us from storing null values in the database, they take different approaches.

As a rule of thumb, we should prefer the @NotNull annotation over the @Column(nullable = false) annotation. This way, we make sure the validation takes place before Hibernate sends any insert or update SQL queries to the database.

Also, it's usually better to rely on the standard rules defined in the Bean Validation, rather than letting the database handle the validation logic.

But, even if we let Hibernate generate the database schema, it'll translate the @NotNull annotation into the database constraints. We must then only make sure that hibernate.validator.apply_to_ddl property is set to true.

MathisHTW replied:
https://www.baeldung.com/hibernate-notnull-vs-nullable

MathisHTW replied:
Nevermind validation before db will take place in service so @NotNull would be redundant.

MathisHTW commented on src/main/java/com/butbetter/storage/model/Address.java:
```diff
@@ -0,0 +1,142 @@
+    @Column(name = "id", insertable = false, updatable = false, nullable = false)
+    private UUID uuid;
+
+    @Column(name = "name", nullable = false)
```
Use @NotBlank instead of nullable = false for all the String attributes of this class.
@NotBlank will not allow null, "" or " ". nullable = false will allow "" and " " as parameters

Ktechen replied:
Validation is part of service

Ktechen replied:
I can added to NotNull

MathisHTW replied:
ok thats a question we have to solve because, nullable = false is basicaly also validation.
Where the validation takes place has to be discussed

MathisHTW commented on src/main/java/com/butbetter/storage/model/ProductInformation.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name="id", insertable = false, updatable = false, nullable = false)
```
I already marked that @Nullable is better performing in the address class. 
Please remove nullable = false here too and replace it with the @NotNull annotation, if it is not a primitive
datatype

Ktechen replied:
Is same as before NotNull is annotation from beanvalidation and Column is annotation from Hibernate.
And better performing is not really true.

notefox commented on src/main/resources/application.properties:
```diff
@@ -1 +1,7 @@
-
+spring.datasource.url=${PS_URL}
```
does this work on heroku?

Ktechen replied:
We must add this in Heroku .env

MathisHTW commented on src/main/java/com/butbetter/storage/model/ProductInformation.java:
```diff
@@ -0,0 +1,100 @@
+    @Column(name = "amount", nullable = false)
+    private int amount;
+
+    @OneToOne
```
What happens when two products have the same address?
Instead use @ManyToOne annotation 
https://thorben-janssen.com/best-practices-many-one-one-many-associations-mappings/

Ktechen replied:
One Product have one address?

Ktechen replied:
fixed

MathisHTW reviewed (changes requested):
Also where my tests at **reeee** 💯

Ktechen commented on src/test/java/com/butbetter/storage/model/AddressTest.java:
```diff
@@ -0,0 +1,22 @@
+	@Test
+	void testNullHashCodeGeneration() {
+		testAddress = new Address("a", "a", "a", "a", "a", null);
+		try {
```
I think a better style is: 
```java
testAddress = new Address("a", "a", "a", "a", "a", null);
assertThrows(Exception.class, () -> {
      testAddress.hashCode();
    });
```

notefox replied:
then make it like that

Ktechen committed:
STOR-8 Change OneToOne to ManyToOne

Ktechen committed:
STOR-21 Refactor the exception handling

Commit:
merge from main

Ktechen committed:
Merge branch 'main' of https://github.com/But-Better/StorageApi into STOR-3-Feature-Products

Commit:
Merge branch 'main' of https://github.com/But-Better/StorageApi into STOR-3-Feature-Products

notefox reviewed (approved):
passt

notefox reviewed (approved):
If it runs on Heroku, push it

Ktechen committed:
STOR-26 add RestController

Ktechen committed:
Merge branch 'STOR-3-Feature-Products' of https://github.com/But-Better/StorageApi into STOR-3-Feature-Products

Ktechen committed:
STOR-7 change type of Column to joinColumn

Ktechen committed:
STOR-26 add ProductInformationNotFoundException for StorageService

Ktechen committed:
STOR-31 add StorageService

Ktechen committed:
Revert "STOR-7 add local application.properties"

This reverts commit cd15d0a7d9090908deb8195a51e277c3e0f05171.

Ktechen committed:
STOR-7 add local application.properties

Ktechen committed:
STOR-7 add local application.properties

Ktechen committed:
Merge branch 'stage' of https://github.com/But-Better/StorageApi into STOR-3-Feature-Products

Ktechen committed:
STOR-31 moved Services

Ktechen committed:
STOR-21 fix test
[ERROR] Failures:
[ERROR]   AddressTest.testNullHashCodeGeneration:16 Expected java.lang.Exception to be thrown, but nothing was thrown.

Ktechen committed:
STOR-8 Change Date datatype to OffsetDateTime

Ktechen committed:
STOR-31 change if constraint for better code style

Ktechen committed:
STOR-21 Add Model Tests

Ktechen committed:
STOR-31 extend ValidateANewProductInformation

Ktechen committed:
STOR-27 add controller tests for create a ProductInformation

Ktechen committed:
STOR-21 refactor ProductInformationTests

Ktechen committed:
STOR-8 add range to amount

Ktechen committed:
STOR-26 change visibility from @ExceptionHandler

Ktechen committed:
STOR-31 add AddressRepository.java to Service

Ktechen committed:
STOR-8 add Tablename

Ktechen committed:
STOR-8 fix Relation between address

Ktechen committed:
STOR-27 add one tests more

Ktechen committed:
STOR-26 add ExceptionHandler

Ktechen committed:
Add script for psql

Ktechen committed:
STOR-20 update application.properties

Ktechen committed:
STOR-20 update application.properties

Ktechen committed:
STOR-20 update application.properties

Ktechen committed:
STOR-26 fix naming of ProductInformation

Ktechen committed:
STOR-27 add Service Tests

Ktechen committed:
Merge branch 'stage' into STOR-3-Feature-Products

Ktechen committed:
STOR-27 disabled systemtests

Ktechen committed:
STOR-8 rename AddressTests

Ktechen committed:
STOR-33 add Postman file

Ktechen committed:
fix merge from pom.xml

notefox commented on src/main/java/com/butbetter/storage/service/StorageService.java:
```diff
@@ -0,0 +1,85 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(StorageService.class);
+
+    private static final String couldNotfoundUser = "User not found with the UUID";
```
If you use the variable once, you don't need to do this up here
just define it when needed

Ktechen committed:
STOR-10 add comments

Ktechen committed:
STOR-31 add comments

notefox commented on src/main/java/com/butbetter/storage/service/StorageService.java:
```diff
@@ -0,0 +1,85 @@
+            throw new NullPointerException(message);
+        }
+
+        if (productInformation.getAmount() <= 0) {
```
Amount can be 0, when we have 0 Products left
[user story](https://butbetter.atlassian.net/browse/STOR-34?atlOrigin=eyJpIjoiMzUzN2M0NjEyODZmNGM4MDg0MTM5Njk1ZjBlNmNkZjQiLCJwIjoiaiJ9)

it would be better to just check, if Amount is negative

Ktechen replied:
Oh :dagger:

Ktechen replied:
fixed

notefox commented on src/test/java/com/butbetter/storage/controller/StorageControllerTests.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @BeforeAll
+    static void beforeAll() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/controller/StorageControllerTests.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @AfterEach
+    void tearDown() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/controller/StorageControllerTests.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @AfterAll
+    static void afterAll() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/model/ProductInformationTests.java:
```diff
@@ -0,0 +1,65 @@
+public class ProductInformationTests {
+
+    @BeforeEach
+    void setUp() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/model/ProductInformationTests.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @BeforeAll
+    static void beforeAll() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/model/ProductInformationTests.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @AfterAll
+    static void afterAll() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/model/ProductInformationTests.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @AfterEach
+    void tearDown() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/service/StorageServiceTests.java:
```diff
@@ -0,0 +1,86 @@
+    private MockMvc mvc;
+
+    @BeforeEach
+    void setUp() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/service/StorageServiceTests.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @BeforeAll
+    static void beforeAll() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

notefox commented on src/test/java/com/butbetter/storage/service/StorageServiceTests.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @AfterAll
+    static void afterAll() {
```
if not needed, it can be removed

Ktechen replied:
auto generated

notefox replied:
if you don't need it, you don't need to generate it

Ktechen committed:
STOR-31 fix 0 is valid

Ktechen committed:
STOR-31 move string in to  Exception

Ktechen committed:
STOR-27 remove empty autogenerate stuff

Ktechen committed:
Add Min and Max test case

Ktechen committed:
Merge branch 'stage' of https://github.com/But-Better/StorageApi into STOR-3-Feature-Products

Ktechen committed:
STOR-20 add Range

Ktechen committed:
Merge branch 'stage' of https://github.com/But-Better/StorageApi into STOR-3-Feature-Products

Ktechen committed:
Add Hibernate Validator
